Import and export `.zip` archive containing assets

somebody1234 opened the pull request:
### Pull Request Description
- Close https://github.com/enso-org/cloud-v2/issues/1943
  - Import a `.zip` archive to be unpacked as separate assets
- Close https://github.com/enso-org/cloud-v2/issues/2096
  - Export a `.zip` archive containing separate assets

### Important Notes
None

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] ~~The documentation has been updated, if necessary.~~
- [x] ~~Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.~~
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [x] ~~Unit tests have been written where possible.~~
- [x] ~~If meaningful changes were made to logic or tests affecting Enso Cloud integration in the libraries, 
      or the Snowflake database integration, a run of the [Extra Tests](https://github.com/enso-org/enso/actions/workflows/extra-nightly-tests.yml) has been scheduled.~~
  - ~~If applicable, it is suggested to paste a link to a successful run of the Extra Tests.~~

somebody1234 committed:
[dashboard] WIP

somebody1234 committed:
[dashboard] Electron endpoint for archive import

somebody1234 committed:
[dashboard/DriveBarToolbar] Initial frontend implementation for archive import

somebody1234 committed:
[ide-desktop] Replace broken `unzip-stream` with `yauzl`

somebody1234 committed:
[ide-desktop] Fix `Content-Length` headers

somebody1234 committed:
[dashboard] Fix CI?

somebody1234 committed:
Merge branch 'develop' into wip/sb/import-archive

somebody1234 committed:
[desktop-client] Switch from `yauzl` to `zip-lib`

somebody1234 committed:
[dashboard] Fix overly eager resetting of directory id

somebody1234 committed:
[dashboard/DriveBarToolbar] Use `filePath` in `importArchive` when available (i.e. on Electron)

somebody1234 committed:
[dashboard/DriveBarToolbar] Disable "import archive" button on Cloud until it is supported

somebody1234 committed:
[desktop-client] Lift up construction of `params`

somebody1234 committed:
[desktop-client] Initial server implementation of `/files/download-archive`

somebody1234 committed:
[dashboard/DriveBarToolbar] Initial UI implementation of "download to archive"

somebody1234 committed:
[dashboard/DriveBarToolbar] Allow picking a file for "download to archive"

somebody1234 committed:
[dashboard/DriveBarToolbar] Disable "Export Archive" button when no assets are selected

somebody1234 committed:
[dashboard/DriveBarToolbar] Show toast when archive is exporting

somebody1234 committed:
[dashboard] Migrate some `LocalStorage` state to `zustand`

somebody1234 committed:
[dashboard] Add persisted `zustand` state for download directory

somebody1234 committed:
[dashboard] Use default download URL when exporting archive

somebody1234 committed:
[dashboard] Carry across old state for local root and additional directories

somebody1234 committed:
[dashboard] Move old `LocalStorage` for local directories to before their first usage

somebody1234 committed:
[dashboard] Fix `localStorage` keys

vitvakatu commented on app/ide-desktop/client/src/preload.ts:
```diff
@@ -201,14 +169,6 @@ exposeInMainWorld(SYSTEM_API_KEY, {
-
 exposeInMainWorld(VERSION_INFO_KEY, debug.VERSION_INFO)
 
-// ==================
-// === MapBox API ===
-// ==================
-
```
Is there any reason removing these comments?

somebody1234 replied:
not really, but is there a reason to keep them, considering that sections are small and usually consist of a single top level syntax element?

vitvakatu commented on app/ide-desktop/client/src/server.ts:
```diff
@@ -303,20 +333,173 @@ export class Server {
-          const url = new URL(`https://example.com/${requestUrl}`)
-          const fileName = url.searchParams.get('file_name')
-          const directory = url.searchParams.get('directory') ?? this.projectsRootDirectory
+    } else if (request.method === 'POST' && request.url?.startsWith('/api/')) {
```
Can we split handlers into separate functions? It is already a 500+ lines long file consisting pretty much of a single switch statement.

somebody1234 committed:
[dashboard] Settings UI to set download folder

somebody1234 committed:
[dashboard] Invalidate Local `listDirectories` when changing local root directory

somebody1234 committed:
[desktop-client] Refactor API functions out of main server class

somebody1234 committed:
[dashboard] Add context menu option for exporting as archive

somebody1234 committed:
[dashboard] `import { type` -> `import type {`

somebody1234 committed:
[dashboard] Fix `exportArchive` request to local server

somebody1234 committed:
[dashboard] Fix exporting archives

somebody1234 committed:
[dashboard] Fix lint errors

somebody1234 committed:
[dashboard] Fix type error

somebody1234 committed:
[dashboard] Developer dark theme

somebody1234 committed:
[dashboard] WIP: Export Archive for Cloud

somebody1234 committed:
[dashboard] Add Cloud support for exporting archive

somebody1234 committed:
[dashboard] Enable 'Export Archive' for Cloud

somebody1234 committed:
Merge branch 'develop' into wip/sb/import-archive

somebody1234 committed:
[dashboard] Use the same URLs for Local and Remote for exporting zip

somebody1234 committed:
Merge branch 'develop' into wip/sb/import-archive

farmaazon commented on app/ide-desktop/client/src/server.ts:
```diff
@@ -88,6 +124,16 @@ async function findPort(port: number): Promise<number> {
+  } catch {
+    return false
+  }
+}
```
Should be not in the "Port Finder" section

farmaazon commented on app/ide-desktop/client/Enso 2025-05-26 19 21 37.zip:
I don't see this file is used anywhere.

somebody1234 replied:
nice catch, thanks

somebody1234 committed:
[desktop-client] Address CR

somebody1234 committed:
[dashboard] Fix TypeScript error

somebody1234 committed:
[desktop-client] Move response handlers to methods

somebody1234 committed:
[dashboard] Use paths for projects on Local Backend

somebody1234 committed:
[desktop-client] Continue implementing conflict resolution on Local backend

somebody1234 committed:
[dashboard] Initial support for archive write streams

somebody1234 committed:
[dashboard] Untested support for streaming writes without going through FS

somebody1234 committed:
[desktop-client] Pack/unpack projects properly to/from `.enso-archive` tar-gzipped archives

somebody1234 committed:
[desktop-client] WIP: Replace project download through Project Manager
with Electron server implementation

somebody1234 committed:
[dashboard] Fix local directory breadcrumbs; [desktop-client] Flatten imported project files

somebody1234 committed:
Merge branch 'develop' into wip/sb/import-archive

somebody1234 committed:
[dashboard] [desktop-client] Begin unifying backends, using Remote backend as source of truth

somebody1234 committed:
[desktop-client] Fix circular import

somebody1234 committed:
Fix import errors

somebody1234 committed:
[dashboard] WIP: Resolve archive conflicts

somebody1234 committed:
Merge branch 'develop' into wip/sb/import-archive

somebody1234 committed:
[dashboard] [desktop-client] (WIP) Electron server support for resolving archive conflicts

somebody1234 committed:
[dashboard] [desktop-client] (WIP) Electron server support for conflict resolution

somebody1234 committed:
[dashboard] [desktop-client] Merge 'upload archive' and 'resolve archive conflicts'

somebody1234 committed:
[dashboard] Show renamed name for archive conflicts

somebody1234 committed:
[dashboard] Polish conflicting archive assets UI

somebody1234 committed:
[dashboard/ConflictingMultipleUploadModal] Show conflicts; filter for only top level assets; strip extensions for display and checking for duplicates

somebody1234 committed:
[dashboard] WIP: Resolve archive conflicts (continued)

somebody1234 committed:
[dashboard] Remove archive conflict resolution

somebody1234 committed:
[dashboard] Rename assets when conflicting when importing archive

TODO: Rename project metadata; archive import endpoint should be same as file upload

somebody1234 committed:
[dashboard] Merge upload project/archive/file endpoints

somebody1234 committed:
[dashboard] [desktop-client] Fix uploading archive

somebody1234 committed:
Prettier

somebody1234 committed:
[dashboard] Fix integration test error

somebody1234 committed:
[dashboard] Wait for archive unzip job to succeed before returning from `uploadFileEnd`

somebody1234 committed:
[dashboard] Fix errors

somebody1234 committed:
[dashboard] `.` should be an invalid asset name

somebody1234 committed:
[bazel] Update digest

somebody1234 committed:
[dashboard] Update developer dark theme

somebody1234 committed:
[dashboard] Fix breadcrumbs display

somebody1234 committed:
[desktop-client/server] Fix renaming conflicting assets

somebody1234 committed:
[ts/Backend] Fix `doesTitleContainInvalidCharacters`

somebody1234 committed:
[ts/Backend] Add more testcases for `doesTitleContainInvalidCharacters`

somebody1234 committed:
[dashboard] Fix crash

somebody1234 committed:
Merge branch 'develop' into wip/sb/import-archive

PabloBuchu commented:
1. Opening project does not work at all. `download-project` throws 404 (works on develop)
2. Can't export to zip a single, selected asset (e.g a directory)
3. `Export archive` does not work in Electron. It sends a request under `localhost:8080`. However it works in the browser so looks like `localBackend` refers to API in browser and PM in Electron <sic!>
4. `Successfully exported archive.` shows up even if the request failed
5. We call `https://localhost:8080/api/download-directory` every time now when I enter a folder in the cloud
6.

somebody1234 committed:
[dashboard] Fix bugs

somebody1234 committed:
Merge branch 'develop' into wip/sb/import-archive

somebody1234 committed:
Fix Bazel lockfile

somebody1234 committed:
[dashboard/Backend] Fix base path

somebody1234 committed:
[dashboard] Fix uploading duplicate projects from zip

somebody1234 committed:
[dashboard] Fix uploading projects

somebody1234 committed:
[dashboard] Update Project Manager middleware shim

somebody1234 committed:
Merge branch 'develop' into wip/sb/import-archive

somebody1234 committed:
[desktop-client/projectManagement] Remove unnecessary exports

somebody1234 committed:
Fix errors

somebody1234 committed:
Fix name of new conflicting project

somebody1234 committed:
[dashboard] Fix incrementing duplicate asset names

somebody1234 committed:
[desktop-client/server] Fix initial path for extract

somebody1234 committed:
[desktop-client/watch] Remove debug code

somebody1234 committed:
[dashboard] Fix export action

somebody1234 committed:
[dashboard] Fix incorrect URL

somebody1234 committed:
[dashboard] Use correct backend to export archive

somebody1234 committed:
[dashboard/RemoteBackend] Export to correct path

somebody1234 committed:
[dashboard] Fix conditions for cloud export

somebody1234 committed:
[dashboard] Avoid loading state for 'export archive' button

somebody1234 committed:
[dashboard] More fixes

somebody1234 committed:
[dashboard] Fixes?

somebody1234 committed:
[dashboard] Add comments

somebody1234 committed:
[dashboard] Fix integration tests

somebody1234 committed:
Merge branch 'develop' into wip/sb/import-archive

somebody1234 committed:
[dashboard] Update `pnpm-lock.yaml`

somebody1234 committed:
[dashboard] Fix import errors

somebody1234 committed:
[dashboard] Fix integration tests

somebody1234 committed:
[dashboard] Prettier

somebody1234 committed:
[desktop-client/server] Fix renaming duplicate assets

somebody1234 committed:
[dashboard/LocalBackend] Throw error when exporting archive fails

somebody1234 committed:
Merge branch 'develop' into wip/sb/import-archive

somebody1234 committed:
Merge branch 'develop' into wip/sb/import-archive

somebody1234 committed:
[ts/Backend] Remove unnecessary `rootPath`

somebody1234 committed:
Merge branch 'develop' into wip/sb/import-archive

somebody1234 committed:
[dashboard] Update `pnpm-lock.yaml`

somebody1234 committed:
Merge branch 'develop' into wip/sb/import-archive

somebody1234 committed:
[dashboard] Fixes

somebody1234 committed:
[dashboard] Fix type error

somebody1234 committed:
Fix import errors

somebody1234 committed:
Merge branch 'develop' into wip/sb/import-archive

somebody1234 commented:
@vitvakatu hey - would be nice if you could do a basic QA of the file drag-n-drop, because it seems like you touched related code recently (and this PR has a merge conflict in that code)

somebody1234 committed:
[dashboard] Fix type errors

somebody1234 committed:
[dashboard] Fix errors

somebody1234 committed:
[dashboard] Fix errors

PabloBuchu commented:
When it hits 404 on get zip jon it does not retry. The FE needs to poll for the download url couple of times I guess with 60sec timout?

somebody1234 committed:
Merge branch 'develop' into wip/sb/import-archive

somebody1234 committed:
[dashboard] Attempt to refetch for json

somebody1234 committed:
[dashboard] Fix integration tests

somebody1234 committed:
Merge branch 'develop' into wip/sb/import-archive

somebody1234 committed:
[dashboard/integration-test] WIP: Fixes

somebody1234 committed:
Merge branch 'develop' into wip/sb/import-archive

somebody1234 committed:
[dashboard/integration-test] Hopefully these fixes will work

somebody1234 committed:
[dashboard/integration-test] Fix type errors

somebody1234 committed:
Merge branch 'develop' into wip/sb/import-archive

somebody1234 committed:
[dashboard/integration-test] Fix test

somebody1234 committed:
Merge branch 'develop' into wip/sb/import-archive

somebody1234 committed:
[dashboard] Update `pnpm-lock.yaml`

somebody1234 committed:
[dashboard] Update `pnpm-lock.yaml`

somebody1234 committed:
Merge branch 'develop' into wip/sb/import-archive
